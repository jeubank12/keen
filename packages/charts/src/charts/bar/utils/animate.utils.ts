import { AnimationProps } from 'framer-motion';
import { Layout } from '@keen.io/ui-core';

import { GroupMode } from '../../../types';

type BarMotionProps = {
  x: number;
  y: number;
  height: number;
  width: number;
  groupMode: GroupMode;
  layout: Layout;
};

export const createBarMotion = ({
  groupMode,
  layout,
  height,
  width,
  x,
  y,
}: BarMotionProps): AnimationProps => {
  const motion = {
    initial: {},
    exit: {
      opacity: 0,
    },
    animate: {
      x,
      y,
      height,
      width,
      opacity: 1,
      transition: {
        duration: 0.4,
      },
    },
  };

  if (groupMode === 'grouped') {
    motion.initial = {
      opacity: 0,

      ...(layout === 'vertical'
        ? {
            x,
            y: height + y,
            height: 0,
          }
        : { y, x, width: 0 }),
    };
  } else {
    motion.initial = {
      opacity: 0,
      x,
      y,
    };
  }

  return motion;
};
