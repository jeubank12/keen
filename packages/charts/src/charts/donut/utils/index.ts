import {
  generateDonutChart,
  LabelsPosition,
  Options as DonutChartOptions,
} from './chart.utils';

export { generateDonutChart, LabelsPosition, DonutChartOptions };
