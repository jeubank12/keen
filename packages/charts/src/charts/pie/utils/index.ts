import {
  generatePieChart,
  LabelsPosition,
  Options as PieChartOptions,
} from './chart.utils';

export { generatePieChart, LabelsPosition, PieChartOptions };
