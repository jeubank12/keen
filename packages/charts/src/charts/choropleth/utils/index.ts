import { fetchMapTopology, MapType } from './map.utils';
import { generateChoropleth } from './chart.utils';

export { fetchMapTopology, generateChoropleth, MapType };
