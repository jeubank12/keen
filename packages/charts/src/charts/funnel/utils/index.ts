import { generateFunnel, calculateStepPoints } from './chart.utils';

export { generateFunnel, calculateStepPoints };
