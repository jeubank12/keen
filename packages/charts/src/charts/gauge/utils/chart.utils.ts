import { scaleLinear } from 'd3-scale';
import { arc, DefaultArcObject } from 'd3-shape';
import { min, sum } from 'd3-array';
import { ColorMode } from '@keen.io/ui-core';

import { convertDegreesToRadians } from '../../../utils/math.utils';
import { calculateColorScale } from '../../../utils/colors.utils';

import {
  ALIGN_STROKE,
  MAX_DEGREES,
  INNER_WIDTH,
  OUTER_WIDTH,
} from '../constants';

import { Dimension, Margins } from '../../../types';

type Options = {
  data: Record<string, any>[];
  valueKey: string;
  dimension: Dimension;
  margins: Margins;
  colors: string[];
  startAngle: number;
  endAngle: number;
  colorMode: ColorMode;
  colorSteps: number;
  minValue: number;
  maxValue: number;
};

export const generateGauge = ({
  data,
  valueKey,
  dimension,
  margins,
  startAngle,
  endAngle,
  minValue,
  maxValue,
  colors,
  colorMode,
  colorSteps,
}: Options) => {
  const { width, height } = dimension;
  const radius = min([
    width / 2 - margins.left - margins.right,
    height / 2 - margins.top - margins.bottom,
  ]);

  const progressValue = sum(data.map(item => item[valueKey]));
  const getColor = calculateColorScale(0, 1, colorMode, colorSteps, colors);

  const arcStartAngle = convertDegreesToRadians(startAngle);
  const arcEndAngle = convertDegreesToRadians(endAngle);

  const arcProperties = {
    startAngle: arcStartAngle,
    endAngle: arcEndAngle,
  } as DefaultArcObject;

  const degreesScale = scaleLinear()
    .domain([minValue, maxValue])
    .range([startAngle, endAngle]);

  const innerArcPath = arc()
    .innerRadius(radius - INNER_WIDTH)
    .outerRadius(radius - OUTER_WIDTH)(arcProperties);

  const maskPath = arc()
    .innerRadius(radius - INNER_WIDTH - ALIGN_STROKE)
    .outerRadius(radius - OUTER_WIDTH)
    .startAngle(convertDegreesToRadians(degreesScale(progressValue)))
    .endAngle(convertDegreesToRadians(endAngle + ALIGN_STROKE))(
    {} as DefaultArcObject
  );

  const outerArcPath = arc()
    .innerRadius(radius - OUTER_WIDTH)
    .outerRadius(radius)(arcProperties);

  const emptySpaceArcPath = arc()
    .innerRadius(radius - OUTER_WIDTH)
    .outerRadius(radius)({
    startAngle: arcEndAngle,
    endAngle: convertDegreesToRadians(MAX_DEGREES - endAngle),
  } as DefaultArcObject);

  return {
    getColor,
    progressValue,
    emptySpaceArcPath,
    maskPath,
    innerArcPath,
    outerArcPath,
    innerArcColor: getColor(progressValue),
  };
};
