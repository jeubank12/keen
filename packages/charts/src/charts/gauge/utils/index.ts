import { generateGauge } from './chart.utils';

export { generateGauge };
