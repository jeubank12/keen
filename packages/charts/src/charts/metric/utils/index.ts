import { generateMetric } from './metric.utils';
import { formatNumber } from './format.utils';

export { generateMetric, formatNumber };
