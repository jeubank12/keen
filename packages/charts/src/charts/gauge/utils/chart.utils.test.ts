import { generateGauge } from './chart.utils';

import { gaugeChart } from '../gauge-chart.fixtures';

describe('@keen.io/charts', () => {
  describe('<GaugeChart /> - utils', () => {
    it('should fallback to default "minimum" value for "auto" minValue', () => {
      const { minimum } = generateGauge(gaugeChart);

      expect(minimum).toMatchInlineSnapshot(`0`);
    });

    it('should fallback to default "maximum" value for "auto" maxValue', () => {
      const { maximum } = generateGauge(gaugeChart);

      expect(maximum).toMatchInlineSnapshot(`100`);
    });

    it('should create "innerArcs" based on angle range', () => {
      const { innerArcs } = generateGauge(gaugeChart);

      expect(innerArcs.length).toMatchInlineSnapshot(`189`);
    });

    it('should calculate progress value for multiple series', () => {
      const { progressValue } = generateGauge({
        ...gaugeChart,
        data: [
          { 'keen.key': 'Result', 'keen.value': 35 },
          { 'keen.key': 'Result', 'keen.value': 70 },
        ],
      });

      expect(progressValue).toEqual(105);
    });
  });
});
