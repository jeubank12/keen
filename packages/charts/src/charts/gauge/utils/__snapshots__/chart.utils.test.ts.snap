// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@keen.io/charts <GaugeChart /> - utils should calculate "path" for arc mask 1`] = `"M39.26508733790656,-7.632359815061794A40,40,0,0,1,39.26508733790656,7.632359815061794L18.650916485505615,3.6253709121543523A19,19,0,0,0,18.650916485505615,-3.6253709121543523Z"`;
